### PR TITLE
Extend PUT endpoint to accept new zip file

### PR DESCRIPTION
**Pull request by mryan321**
Looks noisier than it is - basically refactored the `validateReq` to new file `formdata.go` - and reused in both POST and PUT endpoints.  Plus some test coverage.

### CODE REVIEW

File: api/interactives.go [hunk 1]
```diff
@@ -128,36 +119,37 @@ func (api *API) GetInteractiveMetadataHandler(w http.ResponseWriter, req *http.R
...
-		log.Error(ctx, "Nothing to update", ErrNoMetadata)
-		return
+
+	// 2. Upload file if requested
+	if formDataRequest.FileData != nil {
+		// Check if duplicate SHA exists
+		i, _ := api.mongoDB.GetActiveInteractiveGivenSha(ctx, formDataRequest.Sha)
+		if i != nil {
+			err = fmt.Errorf("archive already exists id (%s) with sha (%s)", i.ID, i.SHA)
+			http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (mryan321) [inline api/interactives.go:140]: im thnking maybe 409 for these id and title dups? conflict?

File: api/interactives.go [hunk 2]
```diff
@@ -212,15 +205,23 @@ func (api *API) UpdateInteractiveHandler(w http.ResponseWriter, req *http.Reques
...
+	// 6. write to DB
 	err = api.mongoDB.UpsertInteractive(ctx, id, updatedModel)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		log.Error(ctx, "Unable to write to DB", err)
 		return
 	}
 
-	WriteJSONBody(updatedModel, w, http.StatusOK)
+	// 7. get updated model
```
Comment (mryan321) [inline api/interactives.go:218]: i think this is important - if we dont update the model in request in expected way for whatever reason we will send back an incorrect model to whats in db. its a shame we cant update and get document in one request :( 

e.g. the update might not be the full interactive - so archive might not come in every update request.

File: api/formdata.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+func NewFormDataRequest(req *http.Request, api *API, attachmentValidator AttachmentValidator) (*FormDataRequest, error) {
+	f := &FormDataRequest{
+		req: req,
+		api: api,
+	}
+	return f, f.Validate(attachmentValidator)
+}
+
+func (f *FormDataRequest) Validate(attachmentValidator AttachmentValidator) error {
```
Comment (nimitsarup) [inline api/formdata.go]: can be a private function ?

File: api/interactives.go [hunk 4]
```diff
@@ -172,38 +164,39 @@ func (api *API) UpdateInteractiveHandler(w http.ResponseWriter, req *http.Reques
...
+		log.Error(ctx, logMsg, ErrCantUpdateSlug)
 		return
 	}
 
-	// 3. prepare updated model
+	// 5. prepare updated model
 	updatedModel := &models.Interactive{
 		ID:            id,
-		Published:     update.Interactive.Published,
+		Published:     formDataRequest.Update.Interactive.Published,
```
Comment (nimitsarup) [inline api/interactives.go]: cosmetic - maybe assign formDataRequest.Update to a variable and use it (will save you using the long form)

**Review by nimitsarup — DISMISSED** [review]

**Review by nimitsarup — APPROVED** [review]

**Pull request merged**